### PR TITLE
conntrack_store: Make CT update event update processing atomic

### PR DESCRIPTION
**Pull request by priyankar-jain**
There are 2 threads which are listening on CT events (one thread on the src IP based events, and the second thread based on the destination IP events).

There is a case in which both threads can receive the same conntrack event (update or delete). (CT entry's src and dst IP are both part of migratable IP given to the process.)

Now during processing of such an update following happens:

1. The current entry is looked up from the conntrack store.
2. A new entry is generated based on the current entry from CT store and the data we got from the event handler.
3. Finally the old/current entry is freed.

Step1, 2 and 3 are atomic amongst themselves but 1,2,3 is not an atomic operation as a whole.

There is a possibility that wrt threads T1 T2, following sequence of operations happens:

T1: 1, 2
T2: 1
T1: 3
T2: 2, 3

So when T2 performs 2, since the entry is already freed by T1 (at step-3) segmentation fault happens when T2 accesses that memory at its step-2.

This commit fixes it by making 1,2,3 an atomic operation as a whole.

### CODE REVIEW

**Comment by priyankar-jain** [comment]
Current UTs are failing because of a kernel issue used by ubuntu image (in github actions). During delete event, CT mark is not being sent which is used as an unique id in our test cases.

Sample logs: https://pipelines.actions.githubusercontent.com/serviceHosts/47e859b1-6123-420f-8578-25ed85de57ff/_apis/pipelines/1/runs/16/signedlogcontent/2?urlExpires=2023-05-16T05%3A35%3A23.1267656Z&urlSigningMethod=HMACV1&urlSignature=oo4HmmACMgLRkZqvqiWnd%2FBCFDaaNM6BuB32KnnJejk%3D

for now, I've manually run the tests locally on my setup and they are passing fine. Once github ubuntu image gets updated, the tests should be passing automatically.
FYI This kernel bug is fixed upstream: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/net/netfilter?id=9f7dd42f0db1dc6915a52d4a8a96ca18dd8cc34e

**Review by jlevon — COMMENTED** [review]
please rebase this into the other commit It hardly seems minor FWIW

**Comment by jlevon** [comment]
What is "your setup"? Does this correspond to actual AHV?
I'm concerned there is no automated testing of this actually happening at any point.
First, we should not leave the test failing: if it can't work in github, mark the test as expected fail. Don't leave things red.

I guess this is unrelated to this particular change though?

File: src/conntrack_store.c [hunk 1]
```diff
@@ -194,22 +194,22 @@ handle_update_event(struct conntrack_store *conn_store,
...
-    res_ct_entry = get_conntrack_entry_from_update(ct_entry, ct);
     if (res_ct_entry == NULL) {
         LOG(WARNING, "%s: received res_ct_entry as NULL", __func__);
-        return;
+        goto finish;
     }
 
-    pthread_mutex_lock(&conn_store->lock);
-    g_hash_table_replace(conn_store->store, key, res_ct_entry);
+    g_hash_table_insert(conn_store->store, key, res_ct_entry);
```
Comment (jlevon) [inline src/conntrack_store.c]: why this change? They apparently both do the same thing (the docs are hilarious)

**Review by jjd27 — APPROVED** [review]
Change looks good to me, to the best of my knowledge.

**Comment by priyankar-jain** [comment]
@jlevon 
> please rebase this into the other commit It hardly seems minor FWIW

I was planning to use squash and merge, so all commits in this PR will go as a single commit given that they are essentially a single change.

> What is "your setup"? Does this correspond to actual AHV?
I'm concerned there is no automated testing of this actually happening at any point.

So I've tried running these UTs (`make check`) on my devvm (centos7 based):
```
➜  ~ uname -a
Linux priyankar-jain 3.10.0-1127.13.1.el7.x86_64 #1 SMP Tue Jun 23 15:46:38 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
As far automated testing goes, this component is not tested in isolation but we do have regression tests which are testing connection tracking migration upon VM migrations. 

> First, we should not leave the test failing: if it can't work in github, mark the test as expected fail. Don't leave things red.

so I am using check lib for writing UTs. afaik it does not have any flags which can achieve this. The only way is to temporary disable these failing tests. Sorry if i misunderstood ur requirement. 

> I guess this is unrelated to this particular change though?

Yes. it is unrelated to my change. I've triggered these tests without my changes and seeing the same behaviour there as well.

**Comment by priyankar-jain** [comment]
Raised: https://github.com/nutanix/conntrack-migrator/pull/11 to fix UTs.

Once #11 is merged, then will rebase this change on top of it.

File: src/conntrack_store.c [hunk 2]
```diff
@@ -194,22 +194,22 @@ handle_update_event(struct conntrack_store *conn_store,
...
             "Treating it as NEW.", __func__);
-        handle_new_event(conn_store, ct);
-        return;
+        res_ct_entry = conntrack_entry_from_nf_conntrack(ct);
+    } else {
+        res_ct_entry = get_conntrack_entry_from_update(ct_entry, ct);
     }
 
-    res_ct_entry = get_conntrack_entry_from_update(ct_entry, ct);
     if (res_ct_entry == NULL) {
```
Comment (huzaifac2022) [inline src/conntrack_store.c:205]: We can avoid the goto from line 207, and keep the below change between 204 and 213,
res_ct_entry = get_conntrack_entry_from_update(ct_entry, ct);
    if (res_ct_entry == NULL) {
        LOG(WARNING, "%s: received res_ct_entry as NULL", __func__);
    } else {
        g_hash_table_insert(conn_store->store, key, res_ct_entry);
    }

    pthread_mutex_unlock(&conn_store->lock);

Comment (priyankar-jain) [reply src/conntrack_store.c:205]: Done. thanks for the suggestion.

File: src/conntrack_store.c (hunk 1 above)
Comment (priyankar-jain) [reply src/conntrack_store.c]: tbh no reason.
when going thru the docs i found they were doing the same thing. So i used`g_hash_table_insert` instead of `g_hash_table_replace` as we were using insert for the NEW event case as well.
lmk if this is fine otherwise i can revert it back to `g_hash_table_replace` as well.

Comment (jlevon) [reply src/conntrack_store.c]: no it's fine, just wondering if there were something more subtle

**Review by jlevon — APPROVED** [review]

**Review by huzaifac2022 — APPROVED** [review]
Thanks for addressing the comments.
